components/fxa-client/src/jwcrypto.rs

```diff
@@ -0,0 +1,261 @@
+    let (encrypted_key, protected_header, iv, ciphertext, auth_tag) = match encryption_params {
+        EncryptionParameters::ECDH_ES {
+            enc,
+            local_key_pair,
+            peer_jwk,
+        } => {
+            let local_public_key = extract_pub_key_jwk(&local_key_pair)?;
+            let protected_header = JweHeader {
+                alg: Algorithm::ECDH_ES,
+                enc,
+                epk: Some(local_public_key),
+                apu: None,
+                apv: None,
+            };
+
+            let secret = derive_shared_secret(&protected_header, local_key_pair, &peer_jwk)?;
+            let encryption_algorithm = match protected_header.enc {
+                EncryptionAlgorithm::A256GCM => &aead::AES_256_GCM,
+            };
+            let sealing_key = aead::SealingKey::new(encryption_algorithm, &secret.as_ref())?;
+            let additional_data = serde_json::to_string(&protected_header)?;
+            let additional_data =
+                base64::encode_config(additional_data.as_bytes(), base64::URL_SAFE_NO_PAD);
+            let additional_data = additional_data.as_bytes();
+            let aad = aead::Aad::from(additional_data);
+            let mut iv: Vec<u8> = vec![0; 12];
+            rand::fill(&mut iv)?;
+            let nonce = aead::Nonce::try_assume_unique_for_key(encryption_algorithm, &iv)?;
+            let mut encrypted = aead::seal(&sealing_key, nonce, aad, data)?;
+
+            let tag_idx = encrypted.len() - ((128 + 7) >> 3);
+            let tag: Vec<u8> = encrypted.drain(tag_idx..).collect();
+            // encrypted is now the data ciphertext.
+
+            (None, protected_header, iv, encrypted, tag)
+        }
+    };
+
+    // Assemble JWE
+    let encrypted_key = encrypted_key
+        .map(|k: Vec<u8>| base64::encode_config(&k, base64::URL_SAFE_NO_PAD))
+        .unwrap_or_default();
+    let header = serde_json::to_string(&protected_header)?;
+    let header = base64::encode_config(header.as_bytes(), base64::URL_SAFE_NO_PAD);
+    let tag = base64::encode_config(auth_tag, base64::URL_SAFE_NO_PAD);
+    let ciphertext = base64::encode_config(ciphertext, base64::URL_SAFE_NO_PAD);
+    let iv = base64::encode_config(iv, base64::URL_SAFE_NO_PAD);
+    Ok(format!(
+        "{}.{}.{}.{}.{}",
+        header, encrypted_key, iv, ciphertext, tag
+    ))
+}
+
+pub fn decrypt_jwe(jwe: &str, encryption_params: DecryptionParameters) -> Result<String> {
+    let segments: Vec<&str> = jwe.split('.').collect();
+
+    let protected_header: JweHeader = serde_json::from_slice(&base64::decode_config(
+        &segments[0],
+        base64::URL_SAFE_NO_PAD,
+    )?)?;
+    let encrypted_key = match segments[1].is_empty() {
+        true => None,
+        false => Some(base64::decode_config(
+            &segments[1],
+            base64::URL_SAFE_NO_PAD,
+        )?),
+    };
+    let iv = base64::decode_config(&segments[2], base64::URL_SAFE_NO_PAD)?;
```

Comment:
This is not relevant to your changes themselves, but a general observation, what are the chances that this function is called with a malformed JWE, causing the `[]` operator to panic?  (and is that okay)

Reply:
I'll add a check.